Source/SwiftLintFramework/Rules/TrailingNewlineRule.swift

```diff
@@ -19,20 +31,37 @@ public struct TrailingNewlineRule: Rule {
         triggeringExamples: [
             "let a = 0",
             "let a = 0\n\n"
+        ],
+        corrections: [
+            "let a = 0": "let a = 0\n",
+            "let a = 0\n\n": "let a = 0\n"
         ]
     )
 
     public func validateFile(file: File) -> [StyleViolation] {
-        let string = file.contents
-        let start = string.endIndex.advancedBy(-2, limit: string.startIndex)
-        let range = Range(start: start, end: string.endIndex)
-        let substring = string[range].utf16
-        let newLineSet = NSCharacterSet.newlineCharacterSet()
-        let slices = substring.split(allowEmptySlices: true) { !newLineSet.characterIsMember($0) }
-
-        guard let slice = slices.last where slice.count != 1 else { return [] }
-
+        if file.contents.trailingNewlineCount() == 1 {
+            return []
+        }
         return [StyleViolation(ruleDescription: self.dynamicType.description,
             location: Location(file: file.path, line: max(file.lines.count, 1)))]
     }
+
+    public func correctFile(file: File) -> [Correction] {
+        guard let count = file.contents.trailingNewlineCount() where count != 1 else {
+            return []
+        }
+        let region = file.regions().filter {
+            $0.contains(Location(file: file.path, line: max(file.lines.count, 1)))
+        }.first
+        if region?.isRuleDisabled(self) == true {
```

Comment:
All rules happen to be checking if they're disabled for regions. I think ideally `correctFile` wouldn't even be called if that were the case. But that's tough because you'd have to codify what part of the file you aimed to fix.